File: Problems/No.20_Valid_Parentheses/valid_parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        
        stack<char> STACK;  // the stack to save (, [, {
        char check;         // to check the top of the stack
        
        // start compute
        for (int i=0; i<s.length(); i++) {
            
            // check whether to push into the stack or pop out
            if (s[i] == '(' or s[i] == '[' or s[i] == '{') STACK.push(s[i]);
            else {
                
                // check if the stack is empty while you are going to pop out
                if (STACK.empty()) return false;
                else {
                    
                    // check whether it is valid or not
                    check = STACK.top();
                    if ( not(s[i] == ')' and check == '(') and
                         not(s[i] == ']' and check == '[') and
                         not(s[i] == '}' and check == '{')) return false;
                    else STACK.pop();
                    
                }
                
            }
            
        }
        
        // check if the stack is empty when you finish iterating s
        return STACK.empty();
        
    }
};

File: Problems/No.20_Valid_Parentheses/valid_parentheses.py
class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        
        # prepare the stack and the valid patterns
        stack = [] 
        check = {"(":")", "{":"}", "[":"]"}
        
        # start compute
        for element in s:
            
            # decide to push in or pop out
            if (element in "({["):
                stack.append(element)
            else:
                
                # check whether the stack is empty, pattern correctness or not
                if (stack == []):
                    return False
                elif (element != check[stack[-1]]):
                    return False
                else:
                    stack.pop()
            
        return stack == []
